### PR TITLE
ROC-3527: Validate postcode format

### PR DESCRIPTION
**Pull request by sabahirfan**
### JIRA link
[ROC-3527](https://tools.hmcts.net/jira/browse/ROC-3527)

### Change description
Add validation for postcode valid format.
Pull request for claim-store-api is: https://github.com/hmcts/cmc-claim-store/pull/323

### Work checklist

- [x] Unit tests added where applicable

### Developer self-QA run statement

- [x] I have clicked through the running application to see if all changes I made actually work.

### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/main/app/forms/models/address.ts [hunk 1]
```diff
@@ -20,7 +21,7 @@ export class ValidationErrors {
   static readonly CITY_NOT_VALID: string = 'The city must be no longer than $constraint1 characters'
 
   static readonly POSTCODE_REQUIRED: string = 'Enter postcode'
-  static readonly POSTCODE_NOT_VALID: string = 'The postcode must be no longer than $constraint1 characters'
+  static readonly POSTCODE_NOT_VALID: string = 'Enter a valid postcode'
```
Comment (yarpo) [inline src/main/app/forms/models/address.ts]: 'Enter a valid postcode

but

'Enter postcode' 

above. Shouldn't be there an "a" as well (or removed in both)?

File: src/main/app/forms/models/address.ts [hunk 2]
```diff
@@ -71,7 +72,7 @@ export class Address implements CompletableTask {
   @IsNotBlank({ message: ValidationErrors.POSTCODE_REQUIRED, groups: ['claimant', 'defendant', 'response'] })
   @IsCountrySupported(Country.all(), { message: ValidationErrors.CLAIMANT_COUNTRY_NOT_SUPPORTED, groups: ['claimant'] })
   @IsCountrySupported(Country.defendantCountries(), { message: ValidationErrors.DEFENDANT_COUNTRY_NOT_SUPPORTED, groups: ['defendant'] })
-  @MaxLength(ValidationConstants.POSTCODE_MAX_LENGTH, {
```
Comment (yarpo) [inline src/main/app/forms/models/address.ts]: have you removed this const from `ValidationConstants`?

File: src/main/app/forms/validation/validators/isValidPostcode.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    if (value === undefined || value === null) {
+      return true
+    }
+    const UK_POSTCODE_REGEX = /^([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9]?[A-Za-z]))))[0-9][A-Za-z]{2})$/
+    const normalised = value.toString().replace(/\s/g,'')
+    return validator.matches(normalised, new RegExp(UK_POSTCODE_REGEX))
+  }
+
+  defaultMessage (args: ValidationArguments) {
+    return 'Enter a valid postcode'
```
Comment (yarpo) [inline src/main/app/forms/validation/validators/isValidPostcode.ts]: To "a" or not "a" ;)

File: src/main/app/forms/models/address.ts (hunk 2 above)
Comment (sabahirfan) [reply src/main/app/forms/models/address.ts]: done

File: src/main/app/forms/models/address.ts (hunk 1 above)
Comment (sabahirfan) [reply src/main/app/forms/models/address.ts]: All error messages where user enter invalid value has `a`. For example,
```
Enter a valid town/city
Enter a valid year
Enter a valid amount owed
Enter a valid first payment date
```

File: src/main/app/forms/validation/validators/isValidPostcode.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  validate (value: any, args?: ValidationArguments): boolean {
+    if (value === undefined || value === null) {
+      return true
+    }
+    const UK_POSTCODE_REGEX = /^([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9]?[A-Za-z]))))[0-9][A-Za-z]{2})$/
+    const normalised = value.toString().replace(/\s/g,'')
+    return validator.matches(normalised, new RegExp(UK_POSTCODE_REGEX))
+  }
+
+  defaultMessage (args: ValidationArguments) {
```
Comment (timja-kainos) [inline src/main/app/forms/validation/validators/isValidPostcode.ts]: why is there a default message and a configured message?

File: src/main/app/forms/models/address.ts (hunk 1 above)
Comment (sabahirfan) [reply src/main/app/forms/models/address.ts]: Discussed with Tarik and Marc, Both error messages are fine as it is.
`Enter a valid postcode` for invalid input and `Enter postcode` for empty.

File: src/main/app/forms/validation/validators/isValidPostcode.ts (hunk 4 above)
Comment (sabahirfan) [reply src/main/app/forms/validation/validators/isValidPostcode.ts]: The configured message will override the default message. I can remove the default if not required.

File: src/main/app/forms/models/address.ts (hunk 1 above)
Comment (yarpo) [reply src/main/app/forms/models/address.ts]: ok, thank you :)

File: src/integration-test/data/test-data.ts [hunk 5]
```diff
@@ -107,7 +107,7 @@ export function createClaimant (type: PartyType): Party {
       line1: '234 Acacia Road',
       line2: 'a really cool place',
       city: 'Edinburgh',
-      postcode: 'EDE 1AC'
```
Comment (timja-kainos) [inline src/integration-test/data/test-data.ts]: any reason for changing this? I've tried it on your preview app and it seems to work fine

Comment (sabahirfan) [reply src/integration-test/data/test-data.ts]: This is an invalid postcode format.

Comment (timja-kainos) [reply src/integration-test/data/test-data.ts]: Seems to pass your current validation?

Comment (sabahirfan) [reply src/integration-test/data/test-data.ts]: `EDE 1AC` was failing validation as invalid format, so have to update this.

File: src/main/app/forms/validation/validators/isValidPostcode.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  ValidationOptions,
+  ValidatorConstraint,
+  ValidatorConstraintInterface,
+  ValidationArguments
+} from 'class-validator'
+
+import * as validator from 'validator'
+
+@ValidatorConstraint()
+export class IsValidPostcodeConstraint implements ValidatorConstraintInterface {
```
[the remark this answers is not in the document]
Comment (sabahirfan) [reply src/main/app/forms/validation/validators/isValidPostcode.ts]: Will extract that as a separate PR.

[the remark this answers is not in the document]
Comment (sabahirfan) [reply src/main/app/forms/validation/validators/isValidPostcode.ts]: https://github.com/hmcts/cmc-validators/pull/2

File: yarn.lock [hunk 7]
```diff
@@ -17,6 +17,14 @@
   version "1.0.1"
   resolved "https://registry.yarnpkg.com/@hmcts/cmc-supported-browsers/-/cmc-supported-browsers-1.0.1.tgz#201d415b7efdf858c77ddccb66bbe9af258c03bc"
 
+"@hmcts/cmc-validators@^0.1.0":
+  version "0.1.0"
+  resolved "https://registry.yarnpkg.com/@hmcts/cmc-validators/-/cmc-validators-0.1.0.tgz#482d6264637d656368f6b209c3026ec6e29c85bc"
+  dependencies:
+    class-validator "^0.8.1"
+    ts-node "^3.3.0"
```
Comment (geek-soft) [inline yarn.lock]: I do not like the fact `ts-node` and `typescript` make the application bundle. We should remove these from dependencies of `@hmcts/cmc-validators` and make them `devDependencies`

File: src/test/app/drafts/models/address.ts [hunk 8]
```diff
@@ -5,17 +5,31 @@ describe('Address', () => {
...
         city: 'Manchester',
-        postcode: 'SW8 4DA'
+        postcode: 'BB12 7NQ'
       }
       const add: Address = new Address().deserialize(input)
       expect(add.isCompleted()).to.equal(true)
     })
 
-    it('should return false when the task is no postcode', () => {
+    it('should return false when the there is no postcode', () => {
```
[the remark this answers is not in the document]
Comment (sabahirfan) [reply src/test/app/drafts/models/address.ts]: done

File: src/test/app/drafts/models/address.ts [hunk 9]
```diff
@@ -5,17 +5,31 @@ describe('Address', () => {
...
       const add: Address = new Address()
       expect(add.isCompleted()).to.equal(false)
     })
+
+    it('should return false when the there is invalid postcode', () => {
```
[the remark this answers is not in the document]
Comment (sabahirfan) [reply src/test/app/drafts/models/address.ts]: done

File: yarn.lock (hunk 7 above)
Comment (sabahirfan) [reply yarn.lock]: will update `cmc-validators`

**Comment by pragnesh-patel** [comment]
Tested on AAT

**Pull request merged**